{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8618c45-44a8-438b-9a5d-aa209ef2bfd4",
   "metadata": {},
   "source": [
    "# Restful API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c79562af-f4ad-487c-ac32-5ed9901cda8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'error_message': 'Something went wrong. Try again.'}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def query_plain(text, url=\"http://bern2.korea.ac.kr/plain\"):\n",
    "    return requests.post(url, json={'text': text}).json()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    text = \"RBD mutations reduced infectivity of spike relative to the WT by around 2-fold (Figure 5D) and was partially rescued by ΔH69/V70.Based on observations of the effect of ΔH69/V70 on spike incorporation in the WT (Figure 3D), we predicted that the mechanism of increased infectivity for ΔH69/V70 in the context of RBD mutations might be similar.Analysis of virions from cell supernatants and cell lysates indeed showed an increased ratio of spike:p24 (Figures 5E and 5F).As observed for the WT, we also observed increased cleaved S2:FL when ΔH69/V70 was present along with the RBD mutants in PVs (Figures 5E and 5G).ΔH69/V70 is required for optimal B.1.1.7 Alpha variant spike S2 incorporation and infectivitySpike ΔH69/V70 in B.1.1.7 enhances spike infectivity(A) Surface representation of the spike homotrimer in open conformation with one upright RBD overlaid with ribbon representation (PDB: 6ZGG), with different monomers shown in black, pale blue, and gold.The deleted residues H69 and V70 and the residues involved in amino acid substitutions (501, 570, 716, 982, and 1118) and the deletion at position 144 are colored red on each monomer and labeled on the monomer with an upright RBD shown in black.Scissors mark the approximate location of an exposed loop (residues 677–688) containing the furin cleavage site and including residue 681, which is absent from the structure.(B) Representative infectivity of B.1.1.7 with replacement of H69 and V70 versus B.1.1.7 containing spike ΔH69/V70 and WT (D614G) spike; single-round infection by luciferase-expressing lentiviruses pseudotyped with SARS-CoV-2 spike protein on HeLa cells transduced with ACE2.The data represent technical triplicates.(C) Fold change of luciferase expression over the replacement of H69V70 in ACE2-transfected and ACE2- and TMPRSS2-transfected HEK293T cells, A459-ACE2/TMPRSS2 cells, and H1299 cells.The data shown are from three independent experiments, each in technical triplicates (one-sample t test).(D) Representative western blot analysis following transfection of HEK293T cells with spike and lentiviral plasmids.Virion loading was normalized for input virus using RT activity.Antibodies against HIV-1 p24 and spike S2 were used with anti-GAPDH as a loading control.(E and F) S2 to FL spike was analyzed by densitometry, and the S2:FL cleavage ratio was calculated for virions (E) and cell lysates (F).\"\n",
    "    print(query_plain(text))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53377fcf-d91a-4ece-adfd-60ef5026fbaa",
   "metadata": {},
   "source": [
    "# Local installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ffc0133b-c866-45a9-bf99-1ba8429eccdf",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'data' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 7\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m requests\u001b[38;5;241m.\u001b[39mpost(url, json\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtext\u001b[39m\u001b[38;5;124m'\u001b[39m: text}, allow_redirects\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\u001b[38;5;241m.\u001b[39mjson()\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;18m__name__\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__main__\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m----> 7\u001b[0m     text \u001b[38;5;241m=\u001b[39m \u001b[43mdata\u001b[49m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtext\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;66;03m# print(query_plain(text))\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'data' is not defined"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "def query_plain(text, url=\"http://localhost:8888/plain\"):\n",
    "    return requests.post(url, json={'text': text}, allow_redirects=True).json()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    # text = data[\"text\"][0]\n",
    "    # print(query_plain(text))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9e39bd82-5c2e-4007-9da3-8e4dd9bdfbd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f9d42ef4-fda9-4f63-bb5f-2dfa266ba8b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv(\"/home/david.yang1/autolit/viriation/data/pipeline_data/paper_flagging_data/0_chunks_labelled.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "67b871d3-e973-48cf-8152-924498514ea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "text = data[\"text\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d22b21d5-89d7-4f4a-b3b4-508e30921275",
   "metadata": {},
   "outputs": [],
   "source": [
    "port = \"http://172.19.5.205:8888/plain\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3bcbc411-c0db-4b1f-bcd4-e89c5b06525a",
   "metadata": {},
   "outputs": [],
   "source": [
    "text = \"The chemokine receptor 1 CXCR-1 (or IL8R-alpha) is a specific receptor for the interleukin 8 (IL-8), which is chemoattractant for neutrophils and has an important role in the inflammatory response. The polymorphism rs2234671 at position Ex2+860G>C of the CXCR1 gene causes a conservative amino acid substitution (S276T). This single nucleotide polymorphism (SNP) seemed to be functional as it was associated with decreased lung cancer risk. Previous studies of our group found association of haplotypes in the IL8 and in the CXCR2 genes with the multifactorial disease chronic periodontitis. In this study we investigated the polymorphism rs2234671 in 395 Brazilian subjects with and without chronic periodontitis. FINDINGS: Similar distribution of the allelic and genotypic frequencies were observed between the groups (p>0.05). CONCLUSIONS: The polymorphism rs2234671 in the CXCR1 gene was not associated with the susceptibility to chronic periodontitis in the studied Brazilian population.\"\n",
    "\n",
    "json1 = query_plain(text, url = \"http://172.19.5.205:8888/plain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1f1bcae9-1664-4947-a684-1ac20b708856",
   "metadata": {},
   "outputs": [],
   "source": [
    "text = data[\"text\"][0]\n",
    "\n",
    "json2 = query_plain(text, url = port)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "de5ec1e9-5fd6-45be-bdd3-f4233fd295f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "total = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4c8d8173-3b77-435c-8184-751e9588e0ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "total.append(json1)\n",
    "total.append(json2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "576ade82-0864-4d6d-a0a2-2b43458bdfdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'annotations': [{'id': ['NCBIGene:3577'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'chemokine receptor 1 CXCR-1',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.8805454969406128,\n",
       "   'span': {'begin': 4, 'end': 31}},\n",
       "  {'id': ['NCBIGene:3577'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'IL8R-alpha',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9964876770973206,\n",
       "   'span': {'begin': 36, 'end': 46}},\n",
       "  {'id': ['NCBIGene:3576'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'interleukin 8',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9440363645553589,\n",
       "   'span': {'begin': 79, 'end': 92}},\n",
       "  {'id': ['NCBIGene:3576'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'IL-8',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9256450533866882,\n",
       "   'span': {'begin': 94, 'end': 98}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'neutrophils',\n",
       "   'obj': 'cell_type',\n",
       "   'prob': 0.8520631194114685,\n",
       "   'span': {'begin': 130, 'end': 141}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'rs2234671',\n",
       "   'obj': 'DNA',\n",
       "   'prob': 0.4692791998386383,\n",
       "   'span': {'begin': 215, 'end': 224}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'rs2234671',\n",
       "   'mutationType': 'SNP',\n",
       "   'normalizedName': 'rs2234671',\n",
       "   'obj': 'mutation',\n",
       "   'prob': None,\n",
       "   'span': {'begin': 215, 'end': 224}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'Ex2',\n",
       "   'obj': 'RNA',\n",
       "   'prob': 0.44072121381759644,\n",
       "   'span': {'begin': 237, 'end': 240}},\n",
       "  {'id': ['NCBIGene:3577'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'CXCR1 gene',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9949538707733154,\n",
       "   'span': {'begin': 255, 'end': 265}},\n",
       "  {'id': ['mesh:D000596'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'amino acid',\n",
       "   'obj': 'drug',\n",
       "   'prob': 0.9966884851455688,\n",
       "   'span': {'begin': 288, 'end': 298}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'S276T',\n",
       "   'mutationType': 'ProteinMutation',\n",
       "   'normalizedName': 'p|SUB|S|276|T;RS#:2234671',\n",
       "   'obj': 'mutation',\n",
       "   'prob': None,\n",
       "   'span': {'begin': 313, 'end': 318}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'single nucleotide polymorphism',\n",
       "   'obj': 'DNA',\n",
       "   'prob': 0.38504478335380554,\n",
       "   'span': {'begin': 326, 'end': 356}},\n",
       "  {'id': ['omim:211980', 'mesh:D008175'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'lung cancer',\n",
       "   'obj': 'disease',\n",
       "   'prob': 0.9995964169502258,\n",
       "   'span': {'begin': 423, 'end': 434}},\n",
       "  {'id': ['NCBIGene:3576'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'IL8',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9823983311653137,\n",
       "   'span': {'begin': 510, 'end': 513}},\n",
       "  {'id': ['NCBIGene:3579'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'CXCR2 genes',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9903212785720825,\n",
       "   'span': {'begin': 525, 'end': 536}},\n",
       "  {'id': ['mesh:D055113'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'chronic periodontitis',\n",
       "   'obj': 'disease',\n",
       "   'prob': 0.9999848008155823,\n",
       "   'span': {'begin': 569, 'end': 590}},\n",
       "  {'id': ['NCBIGene:105369740'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'rs2234671',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.757296621799469,\n",
       "   'span': {'begin': 639, 'end': 648}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'rs2234671',\n",
       "   'mutationType': 'SNP',\n",
       "   'normalizedName': 'rs2234671',\n",
       "   'obj': 'mutation',\n",
       "   'prob': None,\n",
       "   'span': {'begin': 639, 'end': 648}},\n",
       "  {'id': ['mesh:D055113'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'chronic periodontitis',\n",
       "   'obj': 'disease',\n",
       "   'prob': 0.9999970197677612,\n",
       "   'span': {'begin': 692, 'end': 713}},\n",
       "  {'id': ['NCBIGene:105369740'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'rs2234671',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.5196889042854309,\n",
       "   'span': {'begin': 860, 'end': 869}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'rs2234671',\n",
       "   'mutationType': 'SNP',\n",
       "   'normalizedName': 'rs2234671',\n",
       "   'obj': 'mutation',\n",
       "   'prob': None,\n",
       "   'span': {'begin': 860, 'end': 869}},\n",
       "  {'id': ['NCBIGene:3577'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'CXCR1 gene',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.991533637046814,\n",
       "   'span': {'begin': 877, 'end': 887}},\n",
       "  {'id': ['mesh:D055113'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'chronic periodontitis',\n",
       "   'obj': 'disease',\n",
       "   'prob': 0.9999917149543762,\n",
       "   'span': {'begin': 934, 'end': 955}}],\n",
       " 'text': 'The chemokine receptor 1 CXCR-1 (or IL8R-alpha) is a specific receptor for the interleukin 8 (IL-8), which is chemoattractant for neutrophils and has an important role in the inflammatory response. The polymorphism rs2234671 at position Ex2+860G>C of the CXCR1 gene causes a conservative amino acid substitution (S276T). This single nucleotide polymorphism (SNP) seemed to be functional as it was associated with decreased lung cancer risk. Previous studies of our group found association of haplotypes in the IL8 and in the CXCR2 genes with the multifactorial disease chronic periodontitis. In this study we investigated the polymorphism rs2234671 in 395 Brazilian subjects with and without chronic periodontitis. FINDINGS: Similar distribution of the allelic and genotypic frequencies were observed between the groups (p>0.05). CONCLUSIONS: The polymorphism rs2234671 in the CXCR1 gene was not associated with the susceptibility to chronic periodontitis in the studied Brazilian population.',\n",
       " 'timestamp': 'Mon Jul 15 17:03:32 +0000 2024'}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a594eef4-93b4-4566-9e40-9bc376a6befb",
   "metadata": {},
   "outputs": [],
   "source": [
    "text = data[\"text\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1def8379-0bee-4145-97e6-de1f54d930ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'annotations': [{'id': ['NCBIGene:54938'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'SARS-CoV-2 spike deletion',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9273385405540466,\n",
       "   'span': {'begin': 23, 'end': 48}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'SARS-CoV-2',\n",
       "   'obj': 'species',\n",
       "   'prob': 0.8918880820274353,\n",
       "   'span': {'begin': 23, 'end': 33}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'H69/V70',\n",
       "   'obj': 'DNA',\n",
       "   'prob': 0.5724936127662659,\n",
       "   'span': {'begin': 49, 'end': 56}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'SARS-CoV-2',\n",
       "   'obj': 'species',\n",
       "   'prob': 0.7200864553451538,\n",
       "   'span': {'begin': 164, 'end': 174}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'DNA',\n",
       "   'prob': 0.5162000060081482,\n",
       "   'span': {'begin': 182, 'end': 185}},\n",
       "  {'id': ['NCBIGene:340485'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'ACE2 receptor',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9962042570114136,\n",
       "   'span': {'begin': 364, 'end': 377}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'DNA',\n",
       "   'prob': 0.48964551091194153,\n",
       "   'span': {'begin': 438, 'end': 441}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'V70',\n",
       "   'obj': 'DNA',\n",
       "   'prob': 0.47201597690582275,\n",
       "   'span': {'begin': 442, 'end': 445}},\n",
       "  {'id': ['NCBIGene:3166'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9334616661071777,\n",
       "   'span': {'begin': 585, 'end': 588}},\n",
       "  {'id': ['NCBIGene:9398'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'V70',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.954874575138092,\n",
       "   'span': {'begin': 589, 'end': 592}},\n",
       "  {'id': ['NCBIGene:5650'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'spike proteins',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.8009477853775024,\n",
       "   'span': {'begin': 636, 'end': 650}},\n",
       "  {'id': ['NCBIGene:7846'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'Alpha variant B.1',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.8836637139320374,\n",
       "   'span': {'begin': 796, 'end': 813}},\n",
       "  {'id': ['NCBIGene:7846'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'Alpha variant B.1.1.7 spike',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.860276997089386,\n",
       "   'span': {'begin': 945, 'end': 972}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'Wuhan-1 D614G',\n",
       "   'obj': 'cell_line',\n",
       "   'prob': 0.7041093111038208,\n",
       "   'span': {'begin': 1033, 'end': 1046}},\n",
       "  {'id': ['NCBIGene:3166'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.8960279226303101,\n",
       "   'span': {'begin': 1061, 'end': 1064}},\n",
       "  {'id': ['NCBIGene:9398'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'V70',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9220916628837585,\n",
       "   'span': {'begin': 1065, 'end': 1068}},\n",
       "  {'id': ['NCBIGene:3166'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9385126233100891,\n",
       "   'span': {'begin': 1083, 'end': 1086}},\n",
       "  {'id': ['NCBIGene:9398'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'V70',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9107390642166138,\n",
       "   'span': {'begin': 1087, 'end': 1090}},\n",
       "  {'id': ['NCBIGene:9398'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'V70',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9878592491149902,\n",
       "   'span': {'begin': 1268, 'end': 1271}},\n",
       "  {'id': ['NCBIGene:3166'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9944127202033997,\n",
       "   'span': {'begin': 1316, 'end': 1319}},\n",
       "  {'id': ['NCBIGene:9398'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'V70',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9924715161323547,\n",
       "   'span': {'begin': 1320, 'end': 1323}},\n",
       "  {'id': ['NCBIGene:5708', '6187'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'S2',\n",
       "   'obj': 'gene',\n",
       "   'prob': nan,\n",
       "   'span': {'begin': 1342, 'end': 1344}},\n",
       "  {'id': ['NCBIGene:3166'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9946263432502747,\n",
       "   'span': {'begin': 1383, 'end': 1386}},\n",
       "  {'id': ['NCBIGene:9398'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'V70',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9959471225738525,\n",
       "   'span': {'begin': 1387, 'end': 1390}},\n",
       "  {'id': ['NCBIGene:3166'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9936923980712891,\n",
       "   'span': {'begin': 1471, 'end': 1474}},\n",
       "  {'id': ['NCBIGene:9398'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'V70',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9952417612075806,\n",
       "   'span': {'begin': 1475, 'end': 1478}},\n",
       "  {'id': ['NCBIGene:54938'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'SARS-CoV-2 spike',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9687249064445496,\n",
       "   'span': {'begin': 1535, 'end': 1551}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'SARS-CoV-2',\n",
       "   'obj': 'species',\n",
       "   'prob': 0.9039353728294373,\n",
       "   'span': {'begin': 1535, 'end': 1545}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'H69',\n",
       "   'obj': 'DNA',\n",
       "   'prob': 0.8050602078437805,\n",
       "   'span': {'begin': 1552, 'end': 1555}},\n",
       "  {'id': ['NCBIGene:5650'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'spike',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.6493017673492432,\n",
       "   'span': {'begin': 1660, 'end': 1665}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'virions',\n",
       "   'obj': 'cell_type',\n",
       "   'prob': 0.4709782004356384,\n",
       "   'span': {'begin': 1669, 'end': 1676}},\n",
       "  {'id': ['NCBIGene:28905'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'B.1',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.6649690866470337,\n",
       "   'span': {'begin': 1724, 'end': 1727}},\n",
       "  {'id': ['mesh:D045169'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'SARS',\n",
       "   'obj': 'disease',\n",
       "   'prob': 0.9524303674697876,\n",
       "   'span': {'begin': 1871, 'end': 1875}},\n",
       "  {'id': ['CUI-less'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'SARS-CoV-2',\n",
       "   'obj': 'species',\n",
       "   'prob': 0.8636437058448792,\n",
       "   'span': {'begin': 1871, 'end': 1881}},\n",
       "  {'id': ['NCBIGene:2703'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'spike surface glycoprotein',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.6162785291671753,\n",
       "   'span': {'begin': 1883, 'end': 1909}},\n",
       "  {'id': ['NCBIGene:1636'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'human angiotensin-converting enzyme',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9962344169616699,\n",
       "   'span': {'begin': 1924, 'end': 1959}},\n",
       "  {'id': ['NCBITaxon:9606'],\n",
       "   'is_neural_normalized': False,\n",
       "   'mention': 'human',\n",
       "   'obj': 'species',\n",
       "   'prob': 0.991072952747345,\n",
       "   'span': {'begin': 1924, 'end': 1929}},\n",
       "  {'id': ['NCBIGene:710'],\n",
       "   'is_neural_normalized': True,\n",
       "   'mention': 'hACE2',\n",
       "   'obj': 'gene',\n",
       "   'prob': 0.9951404333114624,\n",
       "   'span': {'begin': 1961, 'end': 1966}}],\n",
       " 'text': 'Recurrent emergence of SARS-CoV-2 spike deletion H69/V70 and its role in the Alpha variant B.1.1.7SummaryWe report severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) spike H69/V70 in multiple independent lineages, often occurring after acquisition of receptor binding motif replacements such as N439K and Y453F, known to increase binding affinity to the ACE2 receptor and confer antibody escape.In vitro, we show that, although H69/V70 itself is not an antibody evasion mechanism, it increases infectivity associated with enhanced incorporation of cleaved spike into virions.H69/V70 is able to partially rescue infectivity of spike proteins that have acquired N439K and Y453F escape mutations by increased spike incorporation.In addition, replacement of the H69 and V70 residues in the Alpha variant B.1.1.7 spike (where H69/V70 occurs naturally) impairs spike incorporation and entry efficiency of the B.1.1.7 spike pseudotyped virus.Alpha variant B.1.1.7 spike mediates faster kinetics of cell-cell fusion than wild-type Wuhan-1 D614G, dependent on H69/V70.Therefore, as H69/V70 compensates for immune escape mutations that impair infectivity, continued surveillance for deletions with functional effects is warranted.Graphical abstractHighlightsSpike H69/V70 does not confer escape from antibodiesSpike H69/V70 increases cleaved S2 and spike infectivityB.1.1.7 requires H69/V70 for efficient cleaved spike incorporation and infectivityB.1.1.7 spike requires H69/V70 for rapid syncytium formationMeng et al.report that the SARS-CoV-2 spike H69/V70 has arisen multiple times.The deletion increases entry efficiency associated with increased cleaved spike in virions and can compensate for loss of infectivity.The B.1.1.7 spike requires H69/V70 for efficient cell entry and cell-cell fusion activity.IntroductionSevere acute respiratory syndrome coronavirus 2 (SARS-CoV-2) spike surface glycoprotein engagement of human angiotensin-converting enzyme (hACE2) is essential for virus entry and infection, and the receptor is found in the respiratory and gastrointestinal tracts.',\n",
       " 'timestamp': 'Thu Jul 11 18:52:35 +0000 2024'}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_plain(text, url = \"http://172.19.4.121:8888/plain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b8771fc3-076b-4c9c-b79d-7690f4f6cf1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>text</th>\n",
       "      <th>label</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Recurrent emergence of SARS-CoV-2 spike deleti...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>The deletion increases entry efficiency associ...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>We demonstrate that the deletion is required f...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>The earliest ΔH69/V70 detected on a D614G back...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>Dotted lines separate the association phase fr...</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>All but two sera demonstrated clear titratable...</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>(G) Western blots of virions and cell lysates ...</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>In addition, the increased total and cleaved S...</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>(E) ΔH69/V70 does not alter the virus entry ro...</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>We therefore hypothesized that the increased s...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>(F and G) Densitometric quantification of the ...</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0.1  Unnamed: 0  \\\n",
       "0              0           0   \n",
       "1              1           1   \n",
       "2              2           2   \n",
       "3              3           3   \n",
       "4              4           4   \n",
       "5              5           5   \n",
       "6              6           6   \n",
       "7              7           7   \n",
       "8              8           8   \n",
       "9              9           9   \n",
       "10            10          10   \n",
       "\n",
       "                                                 text  label  position  \n",
       "0   Recurrent emergence of SARS-CoV-2 spike deleti...      1         0  \n",
       "1   The deletion increases entry efficiency associ...      1         1  \n",
       "2   We demonstrate that the deletion is required f...      1         2  \n",
       "3   The earliest ΔH69/V70 detected on a D614G back...      1         3  \n",
       "4   Dotted lines separate the association phase fr...      1         4  \n",
       "5   All but two sera demonstrated clear titratable...      1         5  \n",
       "6   (G) Western blots of virions and cell lysates ...      1         6  \n",
       "7   In addition, the increased total and cleaved S...      1         7  \n",
       "8   (E) ΔH69/V70 does not alter the virus entry ro...      1         8  \n",
       "9   We therefore hypothesized that the increased s...      1         9  \n",
       "10  (F and G) Densitometric quantification of the ...      1        10  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e916e52a-834c-42b1-b258-e9f8884ce0a8",
   "metadata": {},
   "source": [
    "# NER on flagged papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e8c96d8-5509-4dfd-aa2d-1648560cee6b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
